numpy/_core/src/multiarray/convert_datatype.c

```diff
@@ -2021,6 +2060,11 @@ PyArray_AddCastingImplementation(PyBoundArrayMethodObject *meth)
 /**
  * Add a new casting implementation using a PyArrayMethod_Spec.
  *
+ * Using this function outside of module initialization without holding a
+ * critical section on the castingimpls dict may lead to a race to fill the
+ * dict. Use PyArray_GetGastingImpl to lazily register casts at runtime
+ * safely.
```

Comment:
I manually verified that this function is only used during module initialization or inside the critical section I added in this PR

Reply:
Thanks for checking.  I guess we are fine as long as we enforce that casts from/to a new DType must be registered at the DType creation.  (which is the enforced right now, since this is not public API)